File: app/src/main/java/it/cammino/risuscito/PaginaRenderFullScreen.java
package it.cammino.risuscito;

import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.widget.FloatingActionButton;
import android.util.Log;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.mikepenz.community_material_typeface_library.CommunityMaterial;
import com.mikepenz.iconics.IconicsDrawable;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import it.cammino.risuscito.database.RisuscitoDatabase;
import it.cammino.risuscito.database.dao.CantoDao;
import it.cammino.risuscito.database.entities.Canto;
import it.cammino.risuscito.ui.ThemeableActivity;

public class PaginaRenderFullScreen extends ThemeableActivity {

  public static int speedValue;
  public static boolean scrollPlaying;
  public static int idCanto;
  private static String urlCanto;
  private final String TAG = getClass().getCanonicalName();

  @BindView(R.id.fab_fullscreen_off)
  FloatingActionButton fabFullscreen;

  //    private int defaultZoomLevel = 0;
  //    private int defaultScrollX = 0;
  //    private int defaultScrollY = 0;
  @BindView(R.id.cantoView)
  WebView pageView;
  //    private DatabaseCanti listaCanti;
  private Canto currentCanto;
  private Handler mHandler = new Handler();
  final Runnable mScrollDown =
      new Runnable() {
        public void run() {
          try {
            pageView.scrollBy(0, speedValue);
          } catch (NumberFormatException e) {
            pageView.scrollBy(0, 0);
          }

          mHandler.postDelayed(this, 700);
        }
      };
  private LUtils mLUtils;

  @OnClick(R.id.fab_fullscreen_off)
  public void exitFullscreen() {
    saveZoom();
    //    mLUtils.closeActivityWithFadeOut();
  }

  @Override
  public void onCreate(Bundle savedInstanceState) {
    mLUtils = LUtils.getInstance(PaginaRenderFullScreen.this);
    mLUtils.goFullscreen();
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_pagina_render_fullscreen);
    ButterKnife.bind(this);

    //        listaCanti = new DatabaseCanti(this);

    // recupera il numero della pagina da visualizzare dal parametro passato dalla chiamata
    Bundle bundle = this.getIntent().getExtras();
    urlCanto = bundle.getString(Utility.URL_CANTO);
    speedValue = bundle.getInt(Utility.SPEED_VALUE);
    scrollPlaying = bundle.getBoolean(Utility.SCROLL_PLAYING);
    idCanto = bundle.getInt(Utility.ID_CANTO);
    //        Log.i(getClass().toString(), "urlCanto: " + urlCanto);
    //        Log.i(getClass().toString(), "speedValue: " + speedValue);
    //        Log.i(getClass().toString(), "scrollPlaying: " + scrollPlaying);
    //        Log.i(getClass().toString(), "idCanto: " + idCanto);

    //    getSavedZoom();

    IconicsDrawable icon =
        new IconicsDrawable(PaginaRenderFullScreen.this)
            .icon(CommunityMaterial.Icon.cmd_fullscreen_exit)
            .color(Color.WHITE)
            .sizeDp(24)
            .paddingDp(2);
    fabFullscreen.setImageDrawable(icon);
  }

  @Override
  public void onBackPressed() {
    Log.d(TAG, "onBackPressed: ");
    saveZoom();
    //    mLUtils.closeActivityWithFadeOut();
  }

  @Override
  public void onResume() {
    super.onResume();

    pageView.loadUrl(urlCanto);
    //	    Log.i(this.getClass().toString(), "scrollPlaying? " + scrollPlaying);
    if (scrollPlaying) {
      mScrollDown.run();
    }

    WebSettings webSettings = pageView.getSettings();
    webSettings.setUseWideViewPort(true);
    webSettings.setSupportZoom(true);
    webSettings.setLoadWithOverviewMode(true);

    webSettings.setBuiltInZoomControls(true);
    webSettings.setDisplayZoomControls(false);

    //    if (defaultZoomLevel > 0) pageView.setInitialScale(defaultZoomLevel);
    pageView.setWebViewClient(new MyWebViewClient());
  }

  //    @Override
  //    public void onDestroy() {
  //        if (listaCanti != null)
  //            listaCanti.close();
  //        super.onDestroy();
  //    }

  // recupera e setta lo zoom
  //  private void getSavedZoom() {
  //        SQLiteDatabase db = listaCanti.getReadableDatabase();

  //        String query = "SELECT zoom, scroll_x , scroll_y" +
  //                "  FROM ELENCO" +
  //                "  WHERE _id =  " + idCanto;
  //        Cursor cursor = db.rawQuery(query, null);
  //        cursor.moveToFirst();
  //
  //        defaultZoomLevel = cursor.getInt(0);
  //        defaultScrollX = cursor.getInt(1);
  //        defaultScrollY = cursor.getInt(2);
  //
  //        cursor.close();
  //        db.close();
  //  }

  //  @SuppressWarnings("deprecation")
  //  private void saveZoom() {
  //    defaultZoomLevel = (int) (pageView.getScale() * 100);
  //    defaultScrollX = pageView.getScrollX();
  //    defaultScrollY = pageView.getScrollY();
  //
  //    SQLiteDatabase db = listaCanti.getReadableDatabase();
  //
  //    String sql =
  //        "UPDATE ELENCO"
  //            + "  SET zoom = "
  //            + defaultZoomLevel
  //            + " "
  //            + ", scroll_x = "
  //            + defaultScrollX
  //            + " "
  //            + ", scroll_y = "
  //            + defaultScrollY
  //            + " "
  //            + "  WHERE _id =  "
  //            + idCanto;
  //    db.execSQL(sql);
  //    db.close();
  //  }

  @SuppressWarnings("deprecation")
  private void saveZoom() {
    currentCanto.zoom = (int) (pageView.getScale() * 100);
    currentCanto.scrollX = pageView.getScrollX();
    currentCanto.scrollY = pageView.getScrollY();
    new ZoomSaverTask().execute();
  }

  private class MyWebViewClient extends WebViewClient {
    @Override
    public void onPageFinished(WebView view, String url) {
      view.postDelayed(
          new Runnable() {
            @Override
            public void run() {
              //              if (defaultScrollX > 0 || defaultScrollY > 0)
              //                pageView.scrollTo(defaultScrollX, defaultScrollY);
              new ZoomLoaderTask().execute();
            }
            // Delay the scrollTo to make it work
          },
          500);
      super.onPageFinished(view, url);
    }
  }

  private class ZoomSaverTask extends AsyncTask<Void, Void, Integer> {
    @Override
    protected Integer doInBackground(Void... params) {
      CantoDao mDao = RisuscitoDatabase.getInstance(getApplicationContext()).cantoDao();
      mDao.updateCanto(currentCanto);
      return 0;
    }

    @Override
    protected void onPostExecute(Integer integer) {
      super.onPostExecute(integer);
      mLUtils.closeActivityWithFadeOut();
    }
  }

  private class ZoomLoaderTask extends AsyncTask<Void, Void, Integer> {
    @Override
    protected Integer doInBackground(Void... params) {
      CantoDao mDao = RisuscitoDatabase.getInstance(getApplicationContext()).cantoDao();
      currentCanto = mDao.getCantoById(idCanto);
      return 0;
    }

    @Override
    protected void onPostExecute(Integer integer) {
      super.onPostExecute(integer);
      Log.d(TAG, "onPostExecute: " + currentCanto.zoom + " - " + currentCanto.scrollX + " - " + currentCanto.scrollY);
      if (currentCanto.zoom > 0) pageView.setInitialScale(currentCanto.zoom);
      if (currentCanto.scrollX > 0 || currentCanto.scrollY > 0)
        pageView.scrollTo(currentCanto.scrollX, currentCanto.scrollY);
    }
  }
}

File: app/src/main/java/it/cammino/risuscito/PaginaRenderFullScreen.kt
package it.cammino.risuscito

import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.util.Log
import android.view.View
import androidx.activity.addCallback
import androidx.lifecycle.lifecycleScope
import com.blogspot.atifsoftwares.animatoolib.Animatoo
import com.mikepenz.iconics.IconicsDrawable
import com.mikepenz.iconics.typeface.library.community.material.CommunityMaterial
import com.mikepenz.iconics.utils.colorInt
import com.mikepenz.iconics.utils.paddingDp
import com.mikepenz.iconics.utils.sizeDp
import it.cammino.risuscito.database.RisuscitoDatabase
import it.cammino.risuscito.database.entities.Canto
import it.cammino.risuscito.databinding.ActivityPaginaRenderFullscreenBinding
import it.cammino.risuscito.ui.InitialScrollWebClient
import it.cammino.risuscito.ui.ThemeableActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Runnable
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.nio.charset.Charset

class PaginaRenderFullScreen : ThemeableActivity() {

    private var currentCanto: Canto? = null
    var speedValue: Int = 0
    private var scrollPlaying: Boolean = false
    var idCanto: Int = 0
    private lateinit var htmlContent: String
    private val mHandler = Handler(Looper.getMainLooper())
    private val mScrollDown: Runnable = object : Runnable {
        override fun run() {
            try {
                binding.cantoView.scrollBy(0, speedValue)
            } catch (e: NumberFormatException) {
                binding.cantoView.scrollBy(0, 0)
            }

            mHandler.postDelayed(this, 700)
        }
    }
    private lateinit var binding: ActivityPaginaRenderFullscreenBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mViewModel.mLUtils.goFullscreen()
        binding = ActivityPaginaRenderFullscreenBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // recupera il numero della pagina da visualizzare dal parametro passato dalla chiamata
        val bundle = this.intent.extras
        htmlContent = bundle?.getString(Utility.HTML_CONTENT) ?: ""
        speedValue = bundle?.getInt(Utility.SPEED_VALUE) ?: 0
        scrollPlaying = bundle?.getBoolean(Utility.SCROLL_PLAYING) ?: false
        idCanto = bundle?.getInt(Utility.ID_CANTO) ?: 0

        val icon = IconicsDrawable(this, CommunityMaterial.Icon.cmd_fullscreen_exit).apply {
            colorInt = Color.WHITE
            sizeDp = 24
            paddingDp = 2
        }
        binding.fabFullscreenOff.setImageDrawable(icon)
        binding.fabFullscreenOff.setOnClickListener { lifecycleScope.launch { saveZoom() } }

        onBackPressedDispatcher.addCallback(this) {
            onBackPressedAction()
        }

        lifecycleScope.launch { loadWebView() }

    }

    private fun onBackPressedAction() {
        Log.d(TAG, "onBackPressed: ")
        lifecycleScope.launch { saveZoom() }
    }

    private suspend fun saveZoom() {
        @Suppress("DEPRECATION")
        currentCanto?.let {
            it.zoom = (binding.cantoView.scale * 100).toInt()
            it.scrollX = binding.cantoView.scrollX
            it.scrollY = binding.cantoView.scrollY
            Log.d(TAG, "it.id ${it.id} / it.zoom ${it.zoom} / it.scrollX ${it.scrollX} / it.scrollY ${it.scrollY}")
            withContext(lifecycleScope.coroutineContext + Dispatchers.IO) {
                RisuscitoDatabase.getInstance(applicationContext).cantoDao().updateCanto(it)
            }
        }
        finish()
        Animatoo.animateZoom(this)
    }

    private suspend fun loadWebView() {
        val mDao = RisuscitoDatabase.getInstance(this).cantoDao()
        currentCanto = withContext(lifecycleScope.coroutineContext + Dispatchers.IO) {
            mDao.getCantoById(idCanto)
        }

        // fix per crash su android 4.1
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.JELLY_BEAN)
            binding.cantoView.setLayerType(View.LAYER_TYPE_SOFTWARE, null)

        loadContentIntoWebView(htmlContent)

        val webSettings = binding.cantoView.settings
        webSettings.useWideViewPort = true
        webSettings.setSupportZoom(true)
        webSettings.loadWithOverviewMode = true

        webSettings.builtInZoomControls = true
        webSettings.displayZoomControls = false

        currentCanto?.let {
            if (it.zoom > 0) binding.cantoView.setInitialScale(it.zoom)
        }

        binding.cantoView.webViewClient = InitialScrollWebClient(currentCanto)

        if (scrollPlaying)
            mScrollDown.run()

    }

//    private inner class MyWebViewClient : WebViewClient() {
//        override fun onPageFinished(view: WebView, url: String) {
//            view.postDelayed(600) {
//                if ((currentCanto?.scrollX
//                                ?: 0) > 0 || (currentCanto?.scrollY ?: 0) > 0)
//                    view.scrollTo(
//                            currentCanto?.scrollX
//                                    ?: 0, currentCanto?.scrollY ?: 0)
//            }
//            super.onPageFinished(view, url)
//        }
//    }

    private fun loadContentIntoWebView(content: String?) {
        if (!content.isNullOrEmpty()) binding.cantoView.loadData(Base64.encodeToString(
                content.toByteArray(Charset.forName(ECONDING_UTF8)),
                Base64.DEFAULT), DEFAULT_MIME_TYPE, ECONDING_BASE64)
    }

    companion object {
        private val TAG = PaginaRenderFullScreen::class.java.canonicalName
        private const val DEFAULT_MIME_TYPE = "text/html; charset=utf-8"
        private const val ECONDING_UTF8 = "utf-8"
        private const val ECONDING_BASE64 = "base64"
    }
}
